sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -623,8 +624,14 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val eval = child.genCode(ctx)
     val nullSafeCast = nullSafeCastFunction(child.dataType, dataType, ctx)
+
+    // Below the code comment including `eval.value` and `eval.isNull` is a trick. It makes the two
+    // expr values are referred by this code block.
     ev.copy(code = eval.code +
-      castCode(ctx, eval.value, eval.isNull, ev.value, ev.isNull, dataType, nullSafeCast))
+      code"""
+        // Cast from ${eval.value}, ${eval.isNull}
```

Comment:
how to guarantee this is the only one we need to take care?

Reply:
I'd like to clean this part further. The generated codes in `Cast` are tangled now, IMHO.

Reply:
During changing to `Block` in `Cast`, I do think of this again.

`eval` is the only expression for `Cast`, thus it should be the only `ExprValue` input to `Cast`'s code block we need to take care.

Other `ExprValue` like the children of `eval.value` should be taken care in the block 
`eval.code`.

Reply:
I mean, is there other expressions like `Cast`  that we need to manually add reference?

Reply:
I prefer 2. Explicitly specifying the references sounds a bit verbose to me.

Reply:
let's hold it and see what others think about.

Reply:
Yeah. Ok.

Reply:
In this particular case I think we should not use the string interpolator. My preferred end game would be that the `CodeGenerator` functions will just return blocks (or something like that) instead of an opaque strings. That is definitely something we should do in a follow up, can we for now just manually create the block?

That being said, if we are going to pick then I'd strong prefer option 2. I think option 1 is much harder to work with, and is also potentially buggy (what happens if you get the order wrong).

Reply:
I already replaced string to `Block` in `Cast` locally (not committed yet). I'm fine to work with a manual block anyway for now in this case. We can address this in follow-ups.

Indeed, I'm afraid that it's a bit tedious to identify references and fill them in correct order.

Reply:
@hvanhovell @cloud-fan Do you think we should continue to ban `String` in the code block string context (option 2) in a follow-up? This is a quite big PR for now.

Reply:
another +1 for 2